tilings/algorithms/minimal_gridded_perms.py

```diff
@@ -0,0 +1,447 @@
+        """
+        Yield all minimal gridded perms on the tiling.
+
+        If `yield_non_minimal` is True, then it will yield some gridded perms
+        that are non-minimal, found by the initial_gp method. Even though it
+        may not be minimal, this is useful when trying to determine whether or
+        not a tiling is empty.
+        """
+        if not self.requirements:
+            if Obstruction.empty_perm() not in self.obstructions:
+                yield GriddedPerm.empty_perm()
+            return
+        if len(self.requirements) == 1:
+            for req in self.requirements[0]:
+                yield GriddedPerm.from_gridded_perm(req)
+            return
+
+        yield from self.prepare_queue(yield_non_minimal=yield_non_minimal)
+
+        while self.queue:
+            # take the next gridded permutation of the queue, together with the
+            # theoretical counts to create a gridded permutation containing
+            # each of gps.
+            (gp, gps, last_cell, still_localising,
+             last_mindices) = heappop(self.queue)
+            # if gp was one of the initial_gps that satisfied obs/reqs, but
+            # we weren't sure at the time if it was minimal, then now is the
+            # time to check and yield
+            if gp in self.initial_gps_to_auto_yield:
+                # removing this speed up a later check by a tiny bit
+                self.initial_gps_to_auto_yield.remove(gp)
+                if not self.yielded_subgridded_perm(gp):
+                    self.yielded.add(gp)
+                    yield gp
+                # We move onto the next gridded permutation
+                continue
+            # now we try inserting a new point into the cell
+            for (cell, localised) in self.get_cells_to_try(gp, gps, last_cell,
+                                                           still_localising):
+                # The `localised` bool tells us if we inserted into
+                # a cell as it didn't contain the patterns in the
+                # cell. If not, then we update the last cell, to
+                # ensure we only insert into greater or equal cells
+                next_cell = last_cell if localised else cell
+                # Find the bounding box of where the point can be inserted.
+                mindex, maxdex, minval, maxval = gp.get_bounding_box(cell)
+                # We make sure we only insert to the right of where the last
+                # point was inserted into the cell.
+                mindex = max(mindex, last_mindices[cell])
+                for idx, val in product(range(maxdex, mindex - 1, -1),
+                                        range(minval, maxval + 1)):
+                    nextgp = gp.insert_specific_point(cell, idx, val)
+                    # The work_packets_done is used to ensure the same
+                    # task is not processed twice.
+                    key = (nextgp, gps, next_cell)
```

Comment:
I would change the line below to. It would be a bit less nested.
```python 
                    if key in self.work_packets_done:
                        continue
                    self.work_packets_done.add(key)
```

Reply:
Agree, fixed. Also, observed that the attribute was local to this function, so removed it from the class.